packages/generate/src/builders.ts

```diff
@@ -603,15 +622,15 @@ export class CodeBuilder {
 
     let prefix = "";
     if (ref.dir !== this.dir) {
-      const mod = path.posix.basename(ref.dir, path.posix.extname(ref.dir));
-      prefix = `_${mod}`;
+      const namespaceParts = identRef.name.split("::").slice(0, -1);
+      prefix = `_${namespaceParts.join("")}`;
 
       let importPath = path.posix.join(
         path.posix.relative(
           path.posix.dirname(this.dir),
           path.posix.dirname(ref.dir)
         ),
-        mod
+        namespaceParts.at(-1)!
```

Comment:
TIL: [`Array.prototype.at`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at), nice.

Small nit: since we're referencing this and asserting it, it would be helpful to name this thing.

```ts
const mod = namespaceParts.at(-1) as string;
```

Something like that. wdyt?

Reply:
> TIL: [`Array.prototype.at`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at), nice.

Had to double check if it was supported by all Node versions supported here lol


> Small nit: since we're referencing this and asserting it, it would be helpful to name this thing.

Sure. I wasn't sure what to call it..maybe that's why I inlined.
I didn't like mod, because is `mod` == `A::X` or `X`. Maybe `leafModule`?

`!` vs `as string`: I prefer `!` because from my experience it's safer. Safer because you are asserting less. It's only asserting that the type X is _not null_. Vs X is _not null and it's a string_. I can't think of a scenario of the top of my head, but I've had times where TS considers types to sufficiently overlap enough for the typecast be allowed and the error suppressed.

Reply:
> I prefer ! because from my experience it's safer.

That's fair, not a blocker, it just seemed weird to make a not-null assertion on a variable assignment rather than asserting the type directly. More from the reader's perspective than the type checkers perspective. We could also just assign it and move the assertion to where it's used so:

```ts
const mod = namespaceParts.at(-1);

// ...
  mod!
```

> I didn't like mod, because is mod == A::X or X. Maybe leafModule?

Sure! Or what about `moduleFileName`? Anything somewhat obvious here is fine by me, and the whole thing isn't a blocker, just something that made me pause for a second when reading through the code so thought it was worth bikeshedding for a bit on.

Reply:
> We could also just assign it and move the assertion to where it's used so:

That's definitely not my preference. We know it's not nullable, so that's should be the type of the variable (even if it's inferred). Changing type in usage makes me think there are other code paths that could be nullable.
Also it's the closest to `at()` which is what introduces the nullability that we are suppressing.



> Or what about `moduleFileName`?

I like it.